sensors/sensors.go

```diff
@@ -211,45 +220,85 @@ func (lidar Lidar) TimedLidarSensorReading(ctx context.Context) (TimedLidarSenso
 
 	timeRequestedMetadata, ok := md[contextutils.TimeRequestedMetadataKey]
 	if ok {
-		replay = true
+		live = false
 		readingTime, err = time.Parse(time.RFC3339Nano, timeRequestedMetadata[0])
 		if err != nil {
-			msg := "replay sensor timestamp parse RFC3339Nano error"
-			return TimedLidarSensorReadingResponse{}, errors.Wrap(err, msg)
+			return TimedLidarSensorReadingResponse{}, errors.Wrap(err, replayTimestampErrorMessage)
 		}
 	}
-	return TimedLidarSensorReadingResponse{Reading: buf.Bytes(), ReadingTime: readingTime, Replay: replay}, nil
+	return TimedLidarSensorReadingResponse{Reading: buf.Bytes(), ReadingTime: readingTime, Replay: !live}, nil
 }
 
 // TimedIMUSensorReading returns data from the IMU movement sensor and the time the reading is from.
 // IMU Sensors currently do not support replay capabilities.
 func (imu IMU) TimedIMUSensorReading(ctx context.Context) (TimedIMUSensorReadingResponse, error) {
-	replay := false
-	ctxWithMetadata, md := contextutils.ContextWithMetadata(ctx)
-	linAcc, err := imu.imu.LinearAcceleration(ctxWithMetadata, make(map[string]interface{}))
-	if err != nil {
-		msg := "LinearAcceleration error"
-		return TimedIMUSensorReadingResponse{}, errors.Wrap(err, msg)
-	}
-	angVel, err := imu.imu.AngularVelocity(ctxWithMetadata, make(map[string]interface{}))
-	if err != nil {
-		msg := "AngularVelocity error"
-		return TimedIMUSensorReadingResponse{}, errors.Wrap(err, msg)
-	}
-	readingTime := time.Now().UTC()
+	live := true
 
-	timeRequestedMetadata, ok := md[contextutils.TimeRequestedMetadataKey]
-	if ok {
-		replay = true
-		readingTime, err = time.Parse(time.RFC3339Nano, timeRequestedMetadata[0])
-		if err != nil {
-			msg := "replay sensor timestamp parse RFC3339Nano error"
-			return TimedIMUSensorReadingResponse{}, errors.Wrap(err, msg)
+	var timeLinearAcc, timeAngularVel time.Time
+	var linAcc r3.Vector
+	var angVel spatialmath.AngularVelocity
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return TimedIMUSensorReadingResponse{}, nil
+		default:
+			if timeLinearAcc == defaultTime || timeLinearAcc.Sub(timeAngularVel).Milliseconds() < 0 {
```

Comment:
Could you explain this a bit more?

Reply:
Similar to the imu and lidar sensor process what we are doing here is trying to align data. However, in this case, we are creating a single packet and therefore need to search for linear acceleration and angular velocity data whose timestamps are within a certain tolerance of one another (currently defined to be 20 ms).

In order to do this we keep track of the timestamps of the current measurements and re-poll for whichever measurement is before the other, until we have two values within this tolerance.

Reply:
Note: currently I am using a magic number of 20ms for this tolerance. What are your thoughts on me using a value correlated to the imu_data_rate. Say `tolerance = imu_data_rate` or tolerance = `imu_data_rate/2`?